### PR TITLE
Canary app

### PR DESCRIPTION
**Pull request by mjudeikis**
Work in progress for canary app for testing
```release-note
Add canary application to track cluster health from end-user perspective when upgrading
```

### CODE REVIEW

**Comment by kwoodson** [comment]
@mjudeikis @asalkeld,

Is there an advantage to building our own container for this application?  I would imagine it would be as easy to create a deployment from an (nginx, httpd, etc) and return the status.

Any thoughts on that approach?

File: glide.yaml [hunk 1]
```diff
@@ -65,3 +68,6 @@ import:
   version: d69651ed3497faee15a5363a89578e9991f6d5e2
 - package: go.uber.org/zap
   version: v1.8.0
+- package: github.com/beorn7/perks
```
Comment (kwoodson) [inline glide.yaml]: What is this used for?

Comment (mjudeikis) [reply glide.yaml]: :man_shrugging:

Comment (mjudeikis) [reply glide.yaml]: its prometheus dependecy apperently.

**Comment by mjudeikis** [comment]
> Is there an advantage to building our own container for this application? I would imagine it would be as easy to create a deployment from an (nginx, httpd, etc) and return the status.

The main issue is that most of our application is base frameworks only. So we would still need to write custom logic on top of it. In this case, it would be used azure disk and log metrics into it. And expose it via Prometheus metrics. 

So even if we want to do this we need to maintain the application in another language than go

**Comment by mjudeikis** [comment]
@jim-minter ptal
next step is to get metrics shim to keep hitting it, and see how it behaves.

**Comment by Makdaam** [comment]
lgtm

File: pkg/addons/data/Service/openshift-azure-monitoring/canary.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+apiVersion: v1
+kind: Service
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply pkg/addons/data/Service/openshift-azure-monitoring/canary.yaml]: I was thinking about this just about now. Route or LB type service and IP address?

File: cmd/canary/canary.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+const externalMountPoint = "/data"
+
+type ping struct{}
+
+var _ http.Handler = &ping{}
+
+func (p *ping) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	str := fmt.Sprintf("%s/%s : %s. Hostname: %s", r.RemoteAddr, r.URL, r.UserAgent(), os.Getenv("HOSTNAME"))
+	// open existing file
+	f, err := os.OpenFile(path.Join(externalMountPoint, "lastrequest.log"), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0600)
```
Comment (mjudeikis) [inline cmd/canary/canary.go]: This was not appending file. So if we add hostname it would be nice to see when the hostname changes in the log history.

Comment (jim-minter) [reply cmd/canary/canary.go]: the hostname should never change in the log history ;)
don't append - we have no way of cleaning up the file

File: pkg/addons/data/Service/openshift-azure-monitoring/canary.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jim-minter) [reply pkg/addons/data/Service/openshift-azure-monitoring/canary.yaml]: we will probably need to test both in due course

File: pkg/addons/data/StatefulSet.apps/openshift-azure-monitoring/canary.yaml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        readinessProbe:
+          httpGet:
+            path: healthz/ready
+            port: 8080
+            scheme: HTTP
+          initialDelaySeconds: 10
+        volumeMounts:
+        - mountPath: /data
+          name: data
+  volumeClaimTemplates:
```
Comment (jim-minter) [inline pkg/addons/data/StatefulSet.apps/openshift-azure-monitoring/canary.yaml]: please add a node selector for infra nodes

File: pkg/addons/data/Service/openshift-azure-monitoring/canary.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (mjudeikis) [reply pkg/addons/data/Service/openshift-azure-monitoring/canary.yaml]: Added only route for now.

File: cmd/canary/canary.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+		fmt.Fprintf(w, err.Error())
+	} else {
+		fmt.Fprintf(w, str)
+	}
+}
+
+func main() {
+	p := ping{}
+	http.Handle("/", prometheus.InstrumentHandler("webkv", &p))
+	http.Handle("/metrics", promhttp.Handler())
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply cmd/canary/canary.go]: Is up metrics not coming by default?
It is not being scraped by metrics bridge, but we can add ServiceMonitor to scape it.

[the remark this answers is not in the document]
Comment (mjudeikis) [reply cmd/canary/canary.go]: `up{service="canary"}` yes, up is coming by default. we can have it monitored from geneva!

**Comment by mjudeikis** [comment]
@jim-minter ready to merge if you dont have any other comment. re: adding service - lets discuss separately.

File: cmd/canary/canary.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+type ping struct{}
+
+var _ http.Handler = &ping{}
+
+func (p *ping) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	str := fmt.Sprintf("%s/%s : %s. Node: %s", r.RemoteAddr, r.URL, r.UserAgent(), os.Getenv("HOSTNAME"))
+	err := ioutil.WriteFile(path.Join(externalMountPoint, "lastrequest.log"), []byte(str), 0600)
+	if err != nil {
+		fmt.Fprintf(w, err.Error())
```
Comment (jim-minter) [inline cmd/canary/canary.go]: would be better if this case returned a 500 instead of a 200

**Pull request merged**